### PR TITLE
test: test enclosing type for let+ ... and+ ... in expression

### PR DESCRIPTION
**Pull request by Alizter**
This demonstrates that type enclosing has the incorrect behaviour for letop's with multiple ands.

### CODE REVIEW

**Comment by voodoos** [comment]
First diagnostic: The issue is due to the typedtree nodes not respecting location inclusion / disjointness rules. Some locations issues by the parser are also very imprecise.

**Comment by voodoos** [comment]
I have a first improvement: it is now possible to get the type of the patterns !
On the `let+` and `and+` however the whole expression is selected and it's type shown. 
That's coherent but not very informative.

File: src/analysis/type_utils.ml [hunk 1]
```diff
@@ -30,6 +30,10 @@ open Std
 
 module Verbosity = Mconfig.Verbosity
 
+let protect expr =
```
Comment (Alizter) [inline src/analysis/type_utils.ml:33]: What does protect do?

File: tests/test-dirs/type-enclosing/letop2.t/run.t
Comment (Alizter) [inline tests/test-dirs/type-enclosing/letop2.t/run.t:1]: You've removed some of the tests here, are they no longer needed?

File: src/analysis/type_utils.ml (hunk 1 above)
Comment (voodoos) [reply src/analysis/type_utils.ml:33]: To get the "generic" type of the operator Merlin needs to parse it and type it. But `let++` is not valid syntax so adding parenthesis `(let++)` is required for it to parse correctly. The protect function uses a function from compiler libs to determine is the expression requires parenthesizing before parsing.

**Comment by Alizter** [comment]
Thanks @voodoos!

**Pull request merged**